File: HarvestHub/Models/Product.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace HarvestHub.Models
{
    public class Product
    {
        [Key]
        public int ProductId { get; set; }

        [Required]
        [MaxLength(100)]
        public string Name { get; set; } = string.Empty;

        public string? Description { get; set; }

        [Required]
        [Column(TypeName = "decimal(18,2)")]
        public decimal BasePrice { get; set; } // Цена за кг

        [Required]
        public int CurrentStock { get; set; } // Количество в наличии (в кг)

        [MaxLength(20)]
        public string Unit { get; set; } = "кг"; // Единица измерения (по умолчанию кг)

        // Варианты веса для выбора (JSON строка с массивом: [0.5, 1, 1.5, 2, ... 20])
        public string? WeightOptions { get; set; } // JSON: ["0.5", "1", "1.5", "2", ... "20"]

        public bool AllowCustomWeight { get; set; } = true; // Разрешить произвольный вес через галочку

        [Required]
        public int FarmerId { get; set; }

        public int? CategoryId { get; set; }

        [Required]
        [MaxLength(20)]
        public string Status { get; set; } = "Available";

        public DateTime? HarvestDate { get; set; }
        public DateTime? ExpiryDate { get; set; }
        public string? StorageConditions { get; set; }

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

        // Изображение продукта (URL или путь к файлу)
        [MaxLength(1000000)]
        public string? ImageUrl { get; set; }

        // Navigation properties
        public User Farmer { get; set; } = null!;
        public Category? Category { get; set; }
        public ICollection<OrderItem> OrderItems { get; set; } = new List<OrderItem>();
        public ICollection<Review> Reviews { get; set; } = new List<Review>();
        public ICollection<ProductBatch> ProductBatches { get; set; } = new List<ProductBatch>(); // Добавим
        public ICollection<Favorite> Favorites { get; set; } = new List<Favorite>();
    }
}

File: harvesthub-frontend/src/types/product.ts
// src/types/product.ts
import { imagesApi } from '@/api/images.api'
import { ProductDto } from '@/types/backend';

export interface Product {
    id: string;
    name: string;
    description: string;
    price: number;
    discountPrice?: number;
    images: string[];
    categoryId: string;
    categoryName: string;
    farmerId: string;
    farmerName: string;
    stock: number;
    unit?: string;
    rating?: number;
    reviewCount?: number;
    isOrganic?: boolean;
    tags?: string[];
    isAvailable: boolean;
    status: string;
    createdAt?: string;
    updatedAt?: string;
}

// Добавляем export для функции
export const mapDtoToProduct = (dto: ProductDto): Product & { weightOptions?: number[], allowCustomWeight?: boolean } => ({
    id: dto.productId.toString(),
    name: dto.name,
    description: dto.description || '',
    price: dto.basePrice,
    images: dto.imageUrl ? [imagesApi.getImageUrl(dto.imageUrl)] : [],
    categoryId: dto.categoryId.toString(),
    categoryName: dto.categoryName,
    farmerId: dto.farmerId.toString(),
    farmerName: dto.farmerName,
    stock: dto.currentStock,
    unit: (dto.unit && dto.unit !== 'шт' && dto.unit !== 'коробка') ? dto.unit : 'кг', // Заменяем 'шт' и 'коробка' на 'кг'
    isAvailable: dto.status === 'Available' && dto.currentStock > 0,
    status: dto.status,
    createdAt: dto.createdAt,
    weightOptions: dto.weightOptions, // Добавляем варианты веса
    allowCustomWeight: dto.allowCustomWeight, // Добавляем флаг произвольного веса
    // Опциональные поля - можно не указывать, они undefined по умолчанию
});

// Дополнительная функция для карточки товара
export const mapDtoToProductForCard = (dto: ProductDto): Product => ({
    ...mapDtoToProduct(dto),
    unit: 'кг', // Всегда 'кг' для карточки
    rating: 0,
    reviewCount: 0,
    isOrganic: false,
    tags: [],
    updatedAt: dto.createdAt,
});